File: 2nd.py
class HashEntry:
    def __init__(self, key, value):
        self.key = key
        self.value = value


class Dictionary:
    def __init__(self, capacity):
        self.capacity = capacity
        self.table = [[] for _ in range(capacity)]

    def hash_function(self, key):
        # Compute the hash value based on the key
        # You can use any hash function suitable for your keys
        # Here, we use the built-in hash function and modulo operator
        return hash(key) % self.capacity

    def insert(self, key, value):
        index = self.hash_function(key)
        for entry in self.table[index]:
            if entry.key == key:
                # Key already exists, update the value
                entry.value = value
                return
        # Key does not exist, add a new entry
        self.table[index].append(HashEntry(key, value))

    def find(self, key):
        index = self.hash_function(key)
        for entry in self.table[index]:
            if entry.key == key:
                # Key found, return the corresponding value
                return entry.value
        # Key not found, raise an exception
        raise KeyError("Key not found in the dictionary.")

    def delete(self, key):
        index = self.hash_function(key)
        for entry in self.table[index]:
            if entry.key == key:
                # Key found, remove the entry
                self.table[index].remove(entry)
                return
        # Key not found, do nothing or raise an exception


# Create a dictionary with capacity 10
dict = Dictionary(10)

# Insert key-value pairs
dict.insert("apple", 5)
dict.insert("banana", 3)
dict.insert("orange", 7)

# Find values by keys
try:
    value1 = dict.find("apple")
    value2 = dict.find("banana")
    value3 = dict.find("orange")
    print(f"Values: {value1}, {value2}, {value3}")
except KeyError as e:
    print(e)

# Delete a key-value pair
dict.delete("banana")

# Try to find the deleted key (should raise an exception)
try:
    dict.find("banana")
except KeyError as e:
    print(e)

File: 2nd.cpp
#include <iostream>
#include <list>
using namespace std;

// Class representing a key-value pair
class KeyValuePair {
public:
    int key;
    string value;

    KeyValuePair(int k, string v) {
        key = k;
        value = v;
    }
};

// Class representing a Hash Table
class HashTable {
private:
    int tableSize; // Size of the hash table
    list<KeyValuePair>* table; // Array of linked lists

public:
    // Constructor
    HashTable(int size) {
        tableSize = size;
        table = new list<KeyValuePair>[tableSize];
    }

    // Destructor
    ~HashTable() {
        delete[] table;
    }

    // Hash function to compute the index for a given key
    int hashFunction(int key) {
        return key % tableSize;
    }

    // Insert a key-value pair into the dictionary
    void Insert(int key, string value) {
        int index = hashFunction(key);
        KeyValuePair kvp(key, value);

        // Search for existing key in the linked list at the computed index
        for (auto it = table[index].begin(); it != table[index].end(); ++it) {
            if (it->key == key) {
                // Key already exists, update the value
                it->value = value;
                return;
            }
        }

        // Key not found, insert a new key-value pair at the end of the linked list
        table[index].push_back(kvp);
    }

    // Find the value associated with a given key
    string Find(int key) {
        int index = hashFunction(key);

        // Search for the key in the linked list at the computed index
        for (auto it = table[index].begin(); it != table[index].end(); ++it) {
            if (it->key == key) {
                // Key found, return the corresponding value
                return it->value;
            }
        }

        // Key not found
        return "Key not found";
    }

    // Delete a key-value pair from the dictionary
    void Delete(int key) {
        int index = hashFunction(key);

        // Search for the key in the linked list at the computed index
        for (auto it = table[index].begin(); it != table[index].end(); ++it) {
            if (it->key == key) {
                // Key found, erase the key-value pair
                table[index].erase(it);
                return;
            }
        }

        // Key not found, do nothing
    }
};

int main() {
    // Create a HashTable with a size of 10
    HashTable dictionary(10);

    // Insert key-value pairs
    dictionary.Insert(1, "Apple");
    dictionary.Insert(2, "Banana");
    dictionary.Insert(3, "Orange");

    // Find values using keys
    cout << dictionary.Find(2) << endl; // Output: Banana
    cout << dictionary.Find(4) << endl; // Output: Key not found

    // Delete a key-value pair
    dictionary.Delete(1);

    // Find the deleted key
    cout << dictionary.Find(1) << endl; // Output: Key not found

    return 0;
}
